source/src/lib.rs

```diff
@@ -17,938 +17,31 @@
-    };
-
-    Ok(Tree {
-        path: prefix,
-        entries,
-        info,
-    })
-}
-
-/// Provide the [`Revisions`] for the given `identifier`, looking for the
-/// branches as [`BranchType::Remote`].
-///
-/// If there are no branches then this returns `None`.
-///
-/// # Errors
-///
-///   * If we cannot get the branches from the `Browser`
-pub fn remote_revision<P, U>(
-    browser: &Browser,
-    identifier: P,
-    user: U,
-) -> Result<Option<Revisions<P, U>>, Error>
-where
-    P: Clone + ToString,
-{
-    let remote_branches = branches(browser, Some(into_branch_type(Some(identifier.clone()))))?;
-    Ok(
-        NonEmpty::from_vec(remote_branches).map(|branches| Revisions {
-            identifier,
-            user,
-            branches,
-            // TODO(rudolfs): implement remote peer tags once we decide how
-            // https://radicle.community/t/git-tags/214
-            tags: vec![],
-        }),
-    )
-}
-
-/// Provide the [`Revisions`] for the given `identifier`, looking for the
-/// branches as [`BranchType::Local`].
-///
-/// If there are no branches then this returns `None`.
-///
-/// # Errors
-///
-///   * If we cannot get the branches from the `Browser`
-pub fn local_revisions<P, U>(
-    browser: &Browser,
-    identifier: P,
-    user: U,
-) -> Result<Option<Revisions<P, U>>, Error>
-where
-    P: Clone + ToString,
-{
-    let local_branches = branches(browser, Some(BranchType::Local))?;
-    let tags = tags(browser)?;
-    Ok(
-        NonEmpty::from_vec(local_branches).map(|branches| Revisions {
-            identifier,
-            user,
-            branches,
-            tags,
-        }),
-    )
-}
-
-/// Provide the [`Revisions`] of a replicated peer.
-///
-/// Since a replicated peer is one that we have locally, this means that they
-/// have a user profile to build the revision with.
-///
-/// However, if the [`Peer`] does not have any related branches this function
-/// will return `None`.
-///
-/// # Errors
-///
-///   * If we cannot get the branches from the `Browser`
-pub fn revisions<P, U>(
-    browser: &Browser,
-    peer: Category<P, U>,
-) -> Result<Option<Revisions<P, U>>, Error>
-where
-    P: Clone + ToString,
-{
-    match peer {
-        Category::Local { identifier, user } => local_revisions(browser, identifier, user),
-        Category::Remote { identifier, user } => remote_revision(browser, identifier, user),
-    }
-}
-
-/// Turn an `Option<P>` into a [`BranchType`]. If the `P` is present then this
-/// is set as the remote of the `BranchType`. Otherwise, it's local branch.
-#[must_use]
-pub fn into_branch_type<P>(identifier: Option<P>) -> BranchType
-where
-    P: ToString,
-{
-    identifier.map_or(BranchType::Local, |identifier| BranchType::Remote {
-        // We qualify the remotes as the PeerId + heads, otherwise we would grab the tags too.
-        name: Some(format!("{}/heads", identifier.to_string())),
-    })
-}
+pub mod tag;
+pub use tag::{tags, Tag};
```

Comment:
Tag `exports` nothing besides `tags` and `Tag`. To avoid ambiguity and keep the public API small I’d suggest to not publicly expose the `tag` module. This also applies to other modules.

In general, I feel that exposing everything at the crate level and not exposing any modules would be manageable given the size of the crate and would make it more easily digestible.

Reply:
My thoughts are that it gives flexibility of how an upstream caller imports things from `radicle_source`. If they have their own version of `Tag` or `tags` they can import `radicle_source::tag` and so they're now scoped to `tag::Tag` and `tag::tags`. I suppose they could also scope it to `radicle_source::Tag` and `radicle_source::tags` :thinking: 

In the end, there's nothing to hide behind `mod tag` and so there's no need to make it private either.

What are the cons of exposing modules?

Reply:
> If they have their own version of `Tag` or `tags` they can import `radicle_source::tag` and so they're now scoped to `tag::Tag` and `tag::tags`.

That’s possible but likely confusing since now `tag` (from `radicle_source`) is totally disconnected from `Tag` or `tags` which come from a different module. In that case it’s more likely to use `radicle_source::Tag` as you noted.

> What are the cons of exposing modules?

As I mentioned there are two donwsides. One is that it enables two ways to import things. That may not be a huge issue though. It becomes more annoying if some parts of the module are re-exported at the top-level and some parts aren’t and you need to look up which. The second downside is that it makes it harder to grasp the API through the docs. Instead of seeing everything in the crate page I have to go down into the modules. And those items that are re-exported are listed twice in the documentation which leads to more noise.

Another point worth mentioning is that it’s harder to control the public API when the modules are public. If all modules are private then a glance at `src/lib.rs` tells me what is part of the public API. Otherwise it’s easier for code to sneak in to a module that is only intended for internal use and have `pub(crate)` visibility.

Reply:
I think I have a compromise :) Instead of top-level re-exports of these types we organise them into logical modules that re-export them there. So I already introduced `object/*`. For `tag`, `branch`, and `commit` I'm thinking something like `reference`.

How do you feel about this?

Reply:
Some crates have a `prelude` module that exports the most used types, intended to be used like eg. `use radicle_surf::prelude::*` :shrug:

Reply:
Actually, also on this note: I'd like to de-duplicate some of these types anyways because they can just come from `radicle-surf` in the first place.

Reply:
> I think I have a compromise :)

No need for a compromise. Go with what you feel is right! I just wanted to add some perspective. My comments should not prevent this from getting merged.

> Instead of top-level re-exports of these types we organise them into logical modules that re-export them there.

Wouldn’t this make it even harder to track what is exported and from where?

Reply:
Thanks @geigerzaehler, I'll have another pass at this when I de-duplicate the types :)